### PR TITLE
Added 'No resources found' message to logs command

### LINKED ISSUE #808: kubectl logs command with label selector option doesn't print error messages when there are no pod type resources matching labels
**Is this a request for help?** (If yes, you should use our troubleshooting guide and community support channels, see http://kubernetes.io/docs/troubleshooting/.):

**What keywords did you search in Kubernetes issues before filing this one?** (If you have found any duplicates, you should instead reply there.):

is:issue "kubectl logs"

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): BUG REPORT

**Kubernetes version** (use kubectl version):

```json
Client Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.0", GitCommit:"70132b0f130acc0bed193d9ba59dd186f0e634cf", GitTreeState:"clean", BuildDate:"2019-12-13T11:51:44Z", GoVersion:"go1.13.4", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"15", GitVersion:"v1.15.5", GitCommit:"20c265fef0741dd71a66480e35bd69f18351daea", GitTreeState:"clean", BuildDate:"2019-10-15T19:07:57Z", GoVersion:"go1.12.10", Compiler:"gc", Platform:"linux/amd64"}
```

**What happened**:

There are no pod type resources in default namespace.

```
$ kubectl get pods -l test=test
No resources found in default namespace.

$ kubectl logs -l test=test

$ kubectl logs test
Error from server (NotFound): pods "test" not found

$ kubectl logs deployments/test
Error from server (NotFound): deployments.extensions "test" not found
```

**What you expected to happen**:

kubectl logs (POD | TYPE/NAME) command and kubectl get TYPE command with label selector option print error messages when there are no pod type resources matching pod name or type/name.
However, kubectl logs command with label selector option doesn't print error messages when there are no pod type resources matching labels.
I think it should print the error message "No resources found"

**How to reproduce it** (as minimally and precisely as possible):

See the examples above.

### PR DESCRIPTION
**Pull request by cprayer**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:
This PR adds messaging to the kubectl logs command with label selector option so that if no resources are found, it displays a message saying "No resources found in namespace".

**Which issue(s) this PR fixes**:

Fixes kubernetes/kubectl#808

**Special notes for your reviewer**:

It's my first k8s PR. let me know if I missed anything.

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
NONE
```

### CODE REVIEW

File: staging/src/k8s.io/kubectl/pkg/cmd/logs/logs.go
Comment (neolit123) [inline staging/src/k8s.io/kubectl/pkg/cmd/logs/logs.go]: errors should start with lowercase and not end with a period.

Comment (cprayer) [reply staging/src/k8s.io/kubectl/pkg/cmd/logs/logs.go]: @neolit123 

Thank you for your review! I have fixed it now.

**Comment by neolit123** [comment]
the change looks good to me, but i cannot approve this PR.
/assign @adohe
/unassign

**Comment by cprayer** [comment]
@brianpursley 

Hi! I think adohe has no activity since february. can you review or approve this PR?

/assign brianpursley

File: staging/src/k8s.io/kubectl/pkg/cmd/logs/logs.go
Comment (bseenu) [inline staging/src/k8s.io/kubectl/pkg/cmd/logs/logs.go:322]: There is no point of running the validation of the options if there are no resources found this should be bailed out earlier in Complete function itself, without the needing to go through the Validate() and RunLogs() function

-srini

File: staging/src/k8s.io/kubectl/pkg/cmd/logs/logs_test.go
Comment (bseenu) [inline staging/src/k8s.io/kubectl/pkg/cmd/logs/logs_test.go:378]: You are not passing the label selector here, is this a valid test case?

**Comment by bseenu** [comment]
@cprayer I have added the check in the Complete function only where it exits if it does not find any resources with the label selectors - https://github.com/kubernetes/kubernetes/pull/89815 and will not validate the logoptions

Thanks,

-srini

File: staging/src/k8s.io/kubectl/pkg/cmd/logs/logs.go
Comment (cprayer) [reply staging/src/k8s.io/kubectl/pkg/cmd/logs/logs.go:322]: @bseenu
Complete function is for checking required arguments is valid and Validate function is for checking optional flags is valid. These have to precede before the client sends a request.
(I think the logic of retrieving the pod list in Complete function also has to moves to RunLog function.)
Before the client sends a request, it's impossible to check resources exist or not. So it should be in RunLogs function.

File: staging/src/k8s.io/kubectl/pkg/cmd/logs/logs_test.go
Comment (cprayer) [reply staging/src/k8s.io/kubectl/pkg/cmd/logs/logs_test.go:378]: @bseenu 
It covers all cases when kube-api server returns empty resources with no error.

**Comment by cprayer** [comment]
@bseenu 
I already handle this issue. Could you close your PR?

File: staging/src/k8s.io/kubectl/pkg/cmd/logs/logs.go
Comment (bseenu) [reply staging/src/k8s.io/kubectl/pkg/cmd/logs/logs.go:322]: @cprayer i am still not convinced about validating the log options if there are no resources for logs

I would wait to see what others think about this?

Comment (brianpursley) [reply staging/src/k8s.io/kubectl/pkg/cmd/logs/logs.go:322]: @bseenu, I don’t understand, can you explain your thinking further?  

I think the args are valid at this point, it is just that there could be nothing found in the cluster and this check appears to handle that by returning an error instead of succeeding silently.

It seems ok to me, but maybe I am missing something.

File: staging/src/k8s.io/kubectl/pkg/cmd/logs/logs.go
Comment (brianpursley) [inline staging/src/k8s.io/kubectl/pkg/cmd/logs/logs.go:323]: Question because I’m not super familiar with this part, but is this the correct message?

For example, is it possible for the resource to exist but have no logs available, and if so, wouldn’t the message be more accurate to say something like “no logs found for resources in %s namespace”

I’m ok with the message staying this way if it is accurate.  I just wanted to ask the question...

File: staging/src/k8s.io/kubectl/pkg/cmd/logs/logs.go
Comment (bseenu) [reply staging/src/k8s.io/kubectl/pkg/cmd/logs/logs.go:322]: @cprayer I have checked for other commands and see that resourcelist is built in the run function only, Not sure if there is any reasoning for this command to be built in the Complete function ( anyone? ), you can move the resource builder to RunLogs function , if it has to be in Complete function then we can error out there itself

-srini

File: staging/src/k8s.io/kubectl/pkg/cmd/logs/logs.go
Comment (cprayer) [reply staging/src/k8s.io/kubectl/pkg/cmd/logs/logs.go:323]: @brianpursley 
This PR only consider about no resource to exist like get or describe command.

In this case(resource to exist but have no logs available), it doesn't print error message in this PR. 
I've been thinking about resource to exist but have no logs available but I don't know how to deal with follow option. So I don't care about it.

**Comment by brianpursley** [comment]
Just to be sure, I pulled it down and ran it.

Here is my setup:
```
$ kubectl run nginx --image=nginx --restart=Never -l foo=bar
pod/nginx created
```

Current kubectl (foo=bar2 doesn't tell you nothing was found)
```
$ kubectl logs -l foo=bar
$ echo $?
0
$ kubectl logs -l foo=bar2
$ echo $?
0
```

With this PR:
```
$ _output/bin/kubectl logs -l foo=bar
$ echo $?
0
$ _output/bin/kubectl logs -l foo=bar2
error: no resources found in default namespace
$ echo $?
1
```

/lgtm

**Review by brianpursley — APPROVED** [review]

**Comment by cprayer** [comment]
@liggitt 

Can you approve or review this PR?

/assign liggitt

**Comment by liggitt** [comment]
will defer to @seans3 / @soltysh 
/unassign
/assign seans3 soltysh

File: staging/src/k8s.io/kubectl/pkg/cmd/logs/logs.go
Comment (soltysh) [reply staging/src/k8s.io/kubectl/pkg/cmd/logs/logs.go:322]: I agree that the solution proposed in https://github.com/kubernetes/kubernetes/pull/89815 seems better but this one adds a test, which is very important to have. @cprayer @bseenu try to figure out and combine these two PRs together.

**Review by soltysh — CHANGES_REQUESTED** [review]

File: staging/src/k8s.io/kubectl/pkg/cmd/logs/logs.go
Comment (cprayer) [inline staging/src/k8s.io/kubectl/pkg/cmd/logs/logs.go:326]: @soltysh

Look at this, it counts pods in the same way I do. 
Do you think I change from this code to len(o.Object.(*corev1.PodList).Items) too?

Original code

https://github.com/kubernetes/kubernetes/blob/82baa26905c94398a0d19e1b1ecf54eb8acb6029/staging/src/k8s.io/kubectl/pkg/cmd/logs/logs.go#L316-L334

Changed code(Example)
```go
func (o LogsOptions) RunLogs() error {
    podCount := len(o.Object.(*corev1.PodList).Items)
    if o.Follow && podCount > 1 && podCount > o.MaxFollowConcurrency {
        return fmt.Errorf("you are attempting to follow %d log streams, but maximum allowed concurrency is %d, use --max-log-requests to increase the limit", len(requests), o.MaxFollowConcurrency,)
    }

    requests, err := o.LogsForObject(o.RESTClientGetter, o.Object, o.Options, o.GetPodTimeout, o.AllContainers)
    if err != nil {
        return err
    }

    if o.Follow && podCount > 1 {
        return o.parallelConsumeRequest(requests)
    }
    return o.sequentialConsumeRequest(requests)
}
```

**Comment by cprayer** [comment]
@soltysh

Look at this, it counts pods in the same way I do. 
Do you think I change from this code to len(o.Object.(*corev1.PodList).Items) too?

Original code

https://github.com/kubernetes/kubernetes/blob/82baa26905c94398a0d19e1b1ecf54eb8acb6029/staging/src/k8s.io/kubectl/pkg/cmd/logs/logs.go#L316-L334

Changed code(Example)
```go
func (o LogsOptions) RunLogs() error {
    podCount := len(o.Object.(*corev1.PodList).Items)
    if o.Follow && podCount > 1 && podCount > o.MaxFollowConcurrency {
        return fmt.Errorf("you are attempting to follow %d log streams, but maximum allowed concurrency is %d, use --max-log-requests to increase the limit", len(requests), o.MaxFollowConcurrency,)
    }

    requests, err := o.LogsForObject(o.RESTClientGetter, o.Object, o.Options, o.GetPodTimeout, o.AllContainers)
    if err != nil {
        return err
    }

    if o.Follow && podCount > 1 {
        return o.parallelConsumeRequest(requests)
    }
    return o.sequentialConsumeRequest(requests)
}
```

**Comment by soltysh** [comment]
@cprayer the basic rule is to fail fast, here your approach takes longer to achieve the same objective as the alternative. Maybe these are irrelevant ms in the overall execution time, but from code perspective it's easier to grasp and understand.
/lgtm cancel

**Comment by cprayer** [comment]
@soltysh I agree with your opinion to fail fast. So I add a test to the #89815 solution.


as-is(current kubectl version)

```
$ kubectl run nginx --image=nginx --restart=Never -l foo=bar
pod/nginx created

kubectl logs -l foo=bar
/docker-entrypoint.sh: /docker-entrypoint.d/ is not empty, will attempt to perform configuration
/docker-entrypoint.sh: Looking for shell scripts in /docker-entrypoint.d/
/docker-entrypoint.sh: Launching /docker-entrypoint.d/10-listen-on-ipv6-by-default.sh
10-listen-on-ipv6-by-default.sh: Getting the checksum of /etc/nginx/conf.d/default.conf
10-listen-on-ipv6-by-default.sh: Enabled listen on IPv6 in /etc/nginx/conf.d/default.conf
/docker-entrypoint.sh: Launching /docker-entrypoint.d/20-envsubst-on-templates.sh
/docker-entrypoint.sh: Configuration complete; ready for start up

$ echo $?
0

$ kubectl logs -l foo=bar2

$ echo $?
0
```

to-be(67d4a1b PR version)
```
$ ./kubectl run nginx --image=nginx --restart=Never -l foo=bar
pod/nginx created

$ ./kubectl logs -l foo=bar
/docker-entrypoint.sh: /docker-entrypoint.d/ is not empty, will attempt to perform configuration
/docker-entrypoint.sh: Looking for shell scripts in /docker-entrypoint.d/
/docker-entrypoint.sh: Launching /docker-entrypoint.d/10-listen-on-ipv6-by-default.sh
10-listen-on-ipv6-by-default.sh: Getting the checksum of /etc/nginx/conf.d/default.conf
10-listen-on-ipv6-by-default.sh: Enabled listen on IPv6 in /etc/nginx/conf.d/default.conf
/docker-entrypoint.sh: Launching /docker-entrypoint.d/20-envsubst-on-templates.sh
/docker-entrypoint.sh: Configuration complete; ready for start up

$ echo $?
0

$ ./kubectl logs -l foo=bar2
error: no resources found in default namespace

$ echo $?
1
```

File: staging/src/k8s.io/kubectl/pkg/cmd/logs/logs.go
Comment (soltysh) [inline staging/src/k8s.io/kubectl/pkg/cmd/logs/logs.go:279]: It's not an error that you didn't find anything, let's just print this error similarly how `kubectl get` does, see here: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kubectl/pkg/cmd/get/get.go#L585

**Review by soltysh — CHANGES_REQUESTED** [review]

File: staging/src/k8s.io/kubectl/pkg/cmd/logs/logs.go
Comment (cprayer) [reply staging/src/k8s.io/kubectl/pkg/cmd/logs/logs.go]: @soltysh 
Thanks, I have changed it now.

**Review by soltysh — APPROVED** [review]
/lgtm
/approve

**Comment by soltysh** [comment]
This is a minor fix so I'm pushing this over to 1.21
/milestone v1.21

**Pull request merged**
